yjit/src/core.rs

```diff
@@ -2094,9 +2099,14 @@ fn get_num_versions(blockid: BlockId, inlined: bool) -> usize {
                 .version_map
                 .get(insn_idx)
                 .map(|versions| {
-                    versions.iter().filter(|&&version|
-                        Context::decode(unsafe { version.as_ref() }.ctx).inline() == inlined
-                    ).count()
+                    versions.iter().filter(|&&version| {
+                        let version_ctx = Context::decode(unsafe { version.as_ref() }.ctx);
+                        // Inline versions are counted separately towards MAX_INLINE_VERSIONS.
+                        version_ctx.inline() == ctx.inline() &&
+                            // find_block_versions() finds only blocks with compatible reg_mapping,
+                            // so count only versions with compatible reg_mapping.
+                            version_ctx.reg_mapping == ctx.reg_mapping
+                    }).count()
```

Comment:
Important to look at code size since we are increasing the number of versions generated

Reply:
Can you explain, are you making block versions during calls as inline? Are new versions being marked as inline that were not being marked as inline before?

Reply:
I would also like to get a better picture of how this is interacting with the version number limit. It's clearly under counting versions now, but I'm not sure how big an impact that has.

Reply:
> Important to look at code size since we are increasing the number of versions generated

Before (master):

```
compiled_block_count:         70,803
versions_per_block:            1.501
inline_code_size:          7,503,452
outlined_code_size:        6,436,718
code_region_size:         15,413,248
```

After (this PR):

```
compiled_block_count:         75,454
versions_per_block:            1.600
inline_code_size:          8,359,786
outlined_code_size:        6,892,488
code_region_size:         17,068,032
```

For reference, here's the stats without this `&& version_ctx.reg_mapping == ctx.reg_mapping` change but with the rest of the change in this PR:

```
compiled_block_count:         73,872
versions_per_block:            1.588
inline_code_size:          8,158,683
outlined_code_size:        6,744,236
code_region_size:         16,642,048
```

> Can you explain, are you making block versions during calls as inline?

Yes I am when the callee has `Primitive.attr! :inline_block`, e.g. `Array#each`. No for most callees, which are user-defined or don't have `:inline_block` annotation.

> Are new versions being marked as inline that were not being marked as inline before?

No, nothing about inline was changed in this PR. I only added a comment about it since I touched this place.

> I would also like to get a better picture of how this is interacting with the version number limit. It's clearly under counting versions now, but I'm not sure how big an impact that has.

So the reason why I made this change is because, without this change (i.e. removing `&& version_ctx.reg_mapping == ctx.reg_mapping`), `activerecord` would slow down by 6-8%. I looked at the generated code closely, and one of the frequently-called BlockIds had an increased number of versions than before, which reached 5 (or more), due to a different `reg_mapping`. Thus it erased type information from Context, which was `BlockParamProxy`. Erasing `BlockParamProxy` resulted in always falling back to a generic method call in that block.

Since it's above the max versions limit, it creates `ctx.generic_ctx()` to share a block with the generic Context. However, because the `reg_mapping` was incompatible with exiting blocks, it ended up creating a new block that had a new reg_mapping and no type information. Since generating a generic Context doesn't help in this specific situation and `activerecord` slows down too much, I thought generating more `reg_mapping` variations shouldn't reduce the amount of type information.